src/NHSD.BuyingCatalogue.Identity.Api/Services/AgreementConsentService.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Threading.Tasks;
+using IdentityServer4.Events;
+using IdentityServer4.Models;
+using IdentityServer4.Services;
+using NHSD.BuyingCatalogue.Identity.Api.Repositories;
+using NHSD.BuyingCatalogue.Identity.Common.Results;
+
+namespace NHSD.BuyingCatalogue.Identity.Api.Services
+{
+    internal sealed class AgreementConsentService : IAgreementConsentService
+    {
+        private readonly IEventService _eventService;
+        private readonly IIdentityServerInteractionService _interaction;
+        private readonly IScopeRepository _scopeRepository;
+
+        public AgreementConsentService(
+            IEventService eventService,
+            IIdentityServerInteractionService interaction,
+            IScopeRepository scopeRepository)
+        {
+            _eventService = eventService ?? throw new ArgumentNullException(nameof(eventService));
+            _interaction = interaction ?? throw new ArgumentNullException(nameof(interaction));
+            _scopeRepository = scopeRepository ?? throw new ArgumentNullException(nameof(scopeRepository));
+        }
+
+        public async Task<bool> IsValidReturnUrl(Uri returnUrl)
+        {
+            if (returnUrl is null)
+                throw new ArgumentNullException(nameof(returnUrl));
+
+            return await _interaction.GetAuthorizationContextAsync(returnUrl.ToString()) != null;
+        }
+
+        public async Task<Result<Uri>> GrantConsent(Uri returnUrl, string subjectId)
+        {
+            if (returnUrl is null)
+                throw new ArgumentNullException(nameof(returnUrl));
+
+            var context = await _interaction.GetAuthorizationContextAsync(returnUrl.ToString());
+            if (context == null)
+                return Result.Failure<Uri>();
```

Comment:
Unexpected behaviour around returning a failure but not expressing why i.e. no error detail passed into the Result<T> class. Though what you would tell the user of this method I am not sure.

Reply:
As discussed on our call I'll leave this as is on the basis that it would be difficult to know what to display to the user in the case of an error, and an invalid return URL should only occur in a malicious scenario.